src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CQLSessionCache.java

```diff
@@ -0,0 +1,200 @@
+package io.stargate.sgv2.jsonapi.service.cqldriver;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import io.quarkus.security.UnauthorizedException;
+import io.stargate.sgv2.api.common.StargateRequestInfo;
+import io.stargate.sgv2.jsonapi.JsonApiStartUp;
+import io.stargate.sgv2.jsonapi.config.OperationsConfig;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * CQL session cache to reuse the session for the same tenant and token. The cache is configured to
+ * expire after <code>CACHE_TTL_SECONDS</code> of inactivity and to have a maximum size of <code>
+ * CACHE_TTL_SECONDS</code> sessions.
+ */
+@ApplicationScoped
+public class CQLSessionCache {
+  private static final Logger LOGGER = LoggerFactory.getLogger(JsonApiStartUp.class);
+
+  /** Configuration for the JSON API operations. */
+  private final OperationsConfig operationsConfig;
+
+  /** Stargate request info. */
+  @Inject StargateRequestInfo stargateRequestInfo;
+
+  /**
+   * Default tenant to be used when the backend is OSS cassandra and when no tenant is passed in the
+   * request
+   */
+  private static final String DEFAULT_TENANT = "default_tenant";
+  /** CQL username to be used when the backend is AstraDB */
+  private static final String TOKEN = "token";
+
+  /** CQLSession cache. */
+  private final Cache<SessionCacheKey, CqlSession> sessionCache;
+
+  public static final String ASTRA = "astra";
+  public static final String CASSANDRA = "cassandra";
+  /** Default token property name which will be used by the integration tests */
+  @Inject
+  public CQLSessionCache(OperationsConfig operationsConfig) {
+    this.operationsConfig = operationsConfig;
+    sessionCache =
+        Caffeine.newBuilder()
+            .expireAfterAccess(
+                Duration.ofSeconds(operationsConfig.databaseConfig().sessionCacheTtlSeconds()))
+            .maximumSize(operationsConfig.databaseConfig().sessionCacheMaxSize())
+            .evictionListener(
+                (RemovalListener<SessionCacheKey, CqlSession>)
+                    (sessionCacheKey, session, cause) -> {
+                      if (sessionCacheKey != null) {
+                        if (LOGGER.isDebugEnabled()) {
+                          LOGGER.debug(
+                              "Removing session for tenant : {}", sessionCacheKey.tenantId);
+                        }
+                      }
+                      if (session != null) {
+                        session.close();
```

Comment:
Check what happens on race conditions

Reply:
I created an issue for this https://github.com/stargate/jsonapi/issues/645